bokehjs/src/lib/models/layouts/layout_dom.ts

```diff
@@ -156,25 +156,45 @@ export abstract class LayoutDOMView extends PaneView {
     const created = await this.build_child_views()
     const created_children = new Set(created)
 
-    // First remove and then either reattach existing elements or render and
-    // attach new elements, so that the order of children is consistent, while
-    // avoiding expensive re-rendering of existing views.
-    for (const child_view of this.child_views) {
-      child_view.el.remove()
-    }
+    // The newly generated child_views are added to the shadow_el one-by-one
+    // In order to determine the correct ordering we compute the existing
+    // order and then either insert each item before an existing node or append it.
+    // This ensures correct ordering without removing and then re-adding DOM nodes
+    // which can cause issues for certain virtual DOM implementations (e.g. React).
+    const current_views = Array.from(this.shadow_el.children).flatMap(el => {
+      const view = this.child_views.find(view => view.el === el)
+      return view === undefined ? [] : [view]
+    })
 
+    const added = new Set()
     for (const child_view of this.child_views) {
       const is_new = created_children.has(child_view)
+      const target = child_view.rendering_target()
 
-      const target = child_view.rendering_target() ?? this.shadow_el
       if (is_new) {
-        child_view.render_to(target)
+        child_view.render()
+      }
+
+      if (target !== null) {
+        if (!target.contains(child_view.el)) {
+          if (child_view.el.parentNode !== null) {
+            child_view.el.remove()
+          }
```

Comment:
Not needed, if it's already elsewhere in the DOM then it'll be moved on insertion.

```suggestion
```